lib/sync/endpoint.php

```diff
@@ -0,0 +1,210 @@
+<?php
+/**
+ * Adds an http based signaling server for real time collaboration.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Reads the contents of $fd and returns them unserialized.
+ *
+ * @access private
+ * @internal
+ *
+ * @param integer $fd      A file descriptor.
+ *
+ * @return array Unserialized contents of fd.
+ */
+function _gutenberg_get_contents_from_file_descriptor( $fd ) {
+	$contents_raw = stream_get_contents( $fd );
+	$result       = array();
+	if ( $contents_raw ) {
+		$result = unserialize( $contents_raw );
+	}
+	return $result;
+}
+
+
+/**
+ * Makes the file descriptor content of $fd equal to the serialization of content.
+ * Overwrites what was previously in $fd.
+ *
+ * @access private
+ * @internal
+ *
+ * @param integer $fd      A file descriptor.
+ * @param array   $content Content to be serialized and written in a file descriptor.
+ */
+function _gutenberg_save_contents_to_file_descriptor( $fd, $content ) {
+	rewind( $fd );
+	$data = serialize( $content );
+	fwrite( $fd, $data );
+	ftruncate( $fd, strlen( $data ) );
+}
+
+/**
+ * Handles HTTP requests for the gutenberg_signaling_server action.
+ *
+ * @access private
+ * @internal
+ */
+function _gutenberg_wp_ajax_signaling_server() {
+	session_write_close();
+	$subscriber_to_messages_path = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_messages.txt';
+	// Example: array( 2323232121 => array( 'message hello','handshake message' ) ).
+
+	$topics_to_subscribers_path = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'topics_to_subscribers.txt';
+	// Example: array( 'doc1: array( 2323232121 ), 'doc2: array( 2323232123, 2323232121 ) ).
+
+	$subscribers_to_last_connection_path = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_last_connection.txt';
+	// Example: array( 2323232121 => 34343433323(timestamp) ).
+
+	$subscriber_id = $_REQUEST['unique'];
+	if ( ! $subscriber_id ) {
+		die( 'no identifier' );
+	}
+
+	if ( 'GET' === $_SERVER['REQUEST_METHOD'] ) {
+		header( 'Content-Type: text/event-stream' );
+		header( 'Cache-Control: no-cache' );
+		echo 'retry: 3000' . PHP_EOL;
+		$fd = fopen( $subscriber_to_messages_path, 'c+' );
```

Comment:
Another thought I had on file manipulation is that WordPress already has the `WP_Filesystem`, and I'm not sure why we aren't using that.

Reply:
We need to use functions to control access to files like flock, to avoid race conditions. These functions use file descriptors, and we should read and write to the files using the descriptors that were locked. We may consider expanding WP_Filesystem if in the long term, we opt to use files but right now it is not clear what we will do on the long term, we may opt for database or alternatives so for now I guess we can avoid use the base PHP functions and later on decide the final long term solution.